Port firefox webextension

soapdog opened the pull request:
This is a working version for Firefox. The following relevant changes were made:
- Changed `manifest.json` to include fields needed for Firefox
- Changed some checks on `app.js` to use `hasOwnProperty()` 
- Added a branch on `getBudget()` to use `chrome.storage.local` on Firefox instead of `chrome.storage.sync` which is not available.
- Added new `package` and `package:firefox` script entries to `package.json` to help package the WebExtension for Firefox.

soapdog committed:
Porting browser calories to Firefox. Problem with metal.js state change/access

soapdog committed:
Working version. Fixed problem with chrome.storage.sync not being implemented on Firefox.

soapdog committed:
Added routines for packaging extension as xpi for Firefox.

soapdog committed:
Added Firefox info to Readme

miketaylr commented on src/scripts/app.js:
```diff
@@ -22,11 +22,22 @@ class App extends JSXComponent {
-      this.setState({
-        budget: data
+    if (!chrome.storage.hasOwnProperty("sync")) {
+      // Firefox has no sync right now...
```
Here's the tracking bug for that feature, might be good in a comment? 

https://bugzilla.mozilla.org/show_bug.cgi?id=1220494

soapdog replied:
Added and pushed. Thanks for the heads up.

zenorocha replied:
Thanks mike!

soapdog committed:
Added bugzilla URL for chrome.storage.sync implementation to a comment where the branch is happening.

soapdog committed:
fixed options.js because Firefox has no sync yet.

gabrielrcouto commented:
Tested on Firefox 46.0.1 and it's working. 

The only think that is not working here is the loading view (as you can see on the screenshot).

![captura de tela 2016-05-16 as 13 59 35](https://cloud.githubusercontent.com/assets/2197005/15297024/751b3780-1b6e-11e6-9866-36f1c7cd0787.png)

soapdog commented:
@gabrielrcouto 
takes a while to load the measuring view... but the rest of the process worked right?

gabrielrcouto commented:
@soapdog yes, the rest is working 👍

zenorocha commented on src/scripts/app.js:
```diff
@@ -37,12 +49,13 @@ class App extends JSXComponent {
-        if (response.statusCode >= 400) {
+        
+        // Response doesn't contain status code in here.
+        if (!response.hasOwnProperty("html") ) {
```
Why can't we rely on `statusCode`?

soapdog replied:
While debugging that on Firefox by dumping `response` to the console, the `statusCode` property was not present. If it was on the prototype then it is my bad because those things doesn't appear on the log but at least here, at first glance the code appeared broken so I changed it for some property known to be on the object and not on the prototype chain... does this make sense?

soapdog replied:
Also @zenorocha another interesting bit is that `statusCode` will tell you that you had a sucessful HTTP request and not necessarily that your API is the one answering. A captive portal, intranet paranoid firewall, stupid proxy, all of them might try to hijack and redirect this connection which would lead to status codes below 400 (200 or 3xx) and break the extension. **I know this is such an edge case that might never happen**, still if we check for the presence of a known property from the API, we eliminate all potential edge cases and only treat successful results from the API as such.

Am I too paranoid? 🐹

zenorocha replied:
That is a suuuuper edge case. I don't think we should be doing this, `statusCode` makes more sense.

zenorocha commented on package.json:
```diff
@@ -25,7 +26,9 @@
-    "watch:styles": "onchange 'src/styles/*.scss' -- npm run build:styles"
+    "watch:styles": "onchange 'src/styles/*.scss' -- npm run build:styles",
+    "package": "npm run package:firefox",
+    "package:firefox": "cd dist && zip -r ../browser-calories.xpi * && cd .."
```
`zip` is not cross platform. If we do that Windows users will not be able to execute this. Could you add a npm devDependency to do that instead?

soapdog replied:
Will fix and push new commit.

soapdog replied:
Created a new script to package the extension as xpi. A similar script could be created for chrome. Not sure how.

zenorocha commented on package.json:
```diff
@@ -13,6 +13,7 @@
     "babel-preset-metal": "^3.1.0",
     "babelify": "^7.2.0",
     "browserify": "^13.0.0",
+    "metal-component": "^1.0.0-rc.12",
```
Are you using npm 3? This dependency was supposed to be loaded for you

soapdog replied:
I was using 2.x, it complained about this missing dependency, so I added it. I've upgrade to npm 3. Should we leave it for the benefit of those on LTS releases?

zenorocha replied:
Got it

zenorocha commented:
Looks awesome! I added some comments, could you please check them and also remove the `console.log`'s?

soapdog committed:
Removing console.logs (horrible way of debugging, I know... :sad:)

soapdog committed:
Removed dependencies on `cd` and `zip`. Added new script to package extension as Firefox XPI.

soapdog commented:
Added the requested changes and also added line comments above to the issues raised. ;-) anxious for feedback!

zenorocha commented:
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

zenorocha commented:
Ok, so I just tested it and I don't think this is ready to be published yet. 

1) For some reason the loader only shows when the result is completed. The desired behavior is to show the loader as soon as possible.
2) I noticed some weird scroll bars when the result is showed.
3) There are still some console.logs on the code.

Here's a long gif that demonstrates a heavy site being calculated:

![calories](https://cloud.githubusercontent.com/assets/398893/15444926/8963ab06-1eab-11e6-8cae-228880d3810a.gif)

soapdog committed:
Fix for ugly scrollbars in the measuring view.

soapdog committed:
Fix for ugly scrollbars in the measuring view. Added the scrollbars back when viewing the results table.

soapdog committed:
Removed the remaining logs (sorry my bad)

soapdog commented:
@zenorocha,
>1) For some reason the loader only shows when the result is completed. The desired behavior is to show the loader as soon as possible.

The code for the webextension apparently takes a bit long to load on Firefox, the popup is launching before the code is ready that accounts for the delay we see when launching. I believe that is a Firefox bug. WebExtension is a new thing and being heavily worked out

>2) I noticed some weird scroll bars when the result is showed.

Fixed this one.

>3) There are still some console.logs on the code.

Fixed this one as well.

I am adding a gif below to show it working. Be aware that takes a while to calculate UOL calories (hehe). WebExtensions are new on Firefox but improving every day. I would recommend shipping this even with the "popup launch delay bug" if possible as it can benefit its users.

![browser](https://cloud.githubusercontent.com/assets/23247/15484798/42b41f6a-2112-11e6-9bd6-925c584be965.gif)

zenorocha commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
 dist/scripts/popup.js
 dist/styles/popup.css
+*.xpi
+.idea
```
Will remove `.idea` since this should be ignored on `gitignore_global`

zenorocha commented:
Thanks a lot @soapdog! I think we still have some few things to improve but it's already live in the marketplace ;D https://addons.mozilla.org/en-US/firefox/addon/browser-calories/

zenorocha commented:
I also realized that I'm the only responsible for updating versions on browser vendors so there's no need for us to maintain a cross platform zip code.

zenorocha commented:
@soapdog right now when you click to change the default budget it doesn't do anything. Can you send a PR fixing that?

soapdog commented:
@zenorocha will check what was wrong with the options part and fix it on a new PR. Thanks for the hint.

zenorocha commented:
I was able to fix that and some other issues. Feel free to check these commits if you're interested ;)

soapdog commented:
Great work @zenorocha, will check those commits! :)

zenorocha commented:
Firefox extension is taking forever to be published since they introduced some annoying review rules :/
